influxql/ast.go

```diff
@@ -1235,6 +1235,71 @@ func (s *SelectStatement) RewriteFields(ic IteratorCreator) (*SelectStatement, e
 	return other, nil
 }
 
+var (
+	// Exactly matches ^foo$ or ^$
+	exactMatchRegex = regexp.MustCompile(`^\^.*\$$`)
+)
+
+// RewriteRegexExprs rewrites regex conditions to make beter use of the database
+// index.
+//
+// Conditions that can currently be simplified are:
+//
+//     - host ~= /^foo$/ becomes host = 'foo'
+//     - host ~! /^foo$/ becomes host != 'foo'
+//
+// Note: If a backslash (\) is detected in the condition's value, then the
+// regex will *not* be rewritten. In order to support rewriting regexes with
+// escaped characters we need to write a parser, rather than try to match with
+// a regex.
+func (s *SelectStatement) RewriteRegexConditions() {
+	s.Condition = RewriteExpr(s.Condition, func(e Expr) Expr {
+		be, ok := e.(*BinaryExpr)
+		if !ok || (be.Op != EQREGEX && be.Op != NEQREGEX) {
+			// This expression is not a binary condition or doesn't have a
+			// regex based operator.
+			return e
+		}
+
+		// Handle regex-based condition.
+		var matched bool
+		defer func() {
+			// If we rewrote a regex then we'll need to update the condition's
+			// operator.
+			if !matched {
+				return
+			}
+			switch be.Op {
+			case EQREGEX:
+				be.Op = EQ
+			case NEQREGEX:
+				be.Op = NEQ
+			}
+		}()
+
+		rhs := be.RHS.(*RegexLiteral) // This must be a regex.
+		val := rhs.Val.String()
+
+		if strings.ContainsRune(val, '\\') {
```

Comment:
Be aware its intentionally limited to only do the basics, specifically:
- It only deals with anchored i.e /^....$/
- It only deals with captures at the base level i.e. /^(...)$/
- It only deals with concats containing only literals and char classes
- It only deals with alternatives containing literals and concats

So it doesn't deal with all cases as that would open it up to potential issues but here's some that it does deal with
```go
/^1test$/
/^(tester|wibble|blip)$/
/^(1test[1-2][4-5]|wibble[g-hk])$/
/^(2test1|3test2|3test3|3test5|3test7|3test9|3testx)$/ 
````